gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+          const double mu = up / down;
+          if (mu * mu / 4 > maxBeta) {
+            maxIndex = k;
+            maxBeta = mu * mu / 4;
+            break;
+          }
+        }
+      }
+    }
+
+    beta_ = betas[maxIndex];
+  }
+
+  void perturb(int i) {
+    // generate a 0.03*||x_0||_2 as stated in David's paper
+    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::mt19937 generator (seed);
+    std::uniform_real_distribution<double> uniform01(0.0, 1.0);
+
+    int n = m_n_;
+    Vector disturb;
+    disturb.resize(n);
+    disturb.setZero();
+    for (int i =0; i<n; ++i) {
+      disturb(i) = uniform01(generator);
+    }
+    disturb.normalize();
+
+    Vector x0 = ritz_vectors_.col(i);
+    double magnitude = x0.norm();
+    ritz_vectors_.col(i) = x0 + 0.03 * magnitude * disturb;
+  }
+
+  void init(const Vector x0, const Vector x00) {
+    // initialzie ritz eigen values
+    ritz_val_.resize(m_n_);
+    ritz_val_.setZero();
+
+    // initialzie the ritz converged vector
+    ritz_conv_.resize(m_n_);
+    ritz_conv_.setZero();
+
+    // initialzie ritz eigen vectors
+    ritz_vectors_.resize(m_n_, m_n_);
+    ritz_vectors_.setZero();
+    if (accelerated_){
+      ritz_vectors_.col(0) = perform_op(x0, x00);
+      ritz_vectors_.col(1) = perform_op(ritz_vectors_.col(0), x0);
+    } else {
+      ritz_vectors_.col(0) = perform_op(x0);
+      perturb(0);
+    }
+
+    // setting beta
+    if (accelerated_) {
+      Vector init_resid = ritz_vectors_.col(0);
+      const double up = init_resid.transpose() * A_ * init_resid;
+      const double down = init_resid.transpose().dot(init_resid);
+      const double mu =  up/down;
+      beta_ = mu*mu/4;
+      setBeta();
+    }
+    
+  }
+
+  void init() {
+    Vector x0;
+    Vector x00;
+    if (!S_.isZero(0)) {
+      x0 = S_.row(1);
+      x00 = S_.row(0);
+    } else {
+      x0 = random_vec(m_n_);
+      x00 = random_vec(m_n_);
+    }
+    init(x0, x00);
+  }
+
+  bool converged(double tol, int i) {
+    Vector x = ritz_vectors_.col(i);
+    double theta = x.transpose() * A_ * x;
+
+    // store the ritz eigen value
+    ritz_val_(i) = theta;
+
+    // update beta
+    if (accelerated_) beta_ = std::max(beta_, theta * theta / 4);
+
+    Vector diff = A_ * x - theta * x;
+    double error = diff.lpNorm<1>();
+    if (error < tol) ritz_conv_(i) = 1;
+    return error < tol;
+  }
+
+  int num_converged(double tol) {
+    int num_converge = 0;
+    for (int i=0; i<ritz_vectors_.cols(); i++) {
+      if (converged(tol, i)) {
+        num_converge += 1;
+      }
+      if (!accelerated_ && i<ritz_vectors_.cols()-1) {
+        ritz_vectors_.col(i+1) = perform_op(i+1);
+      } else if (accelerated_ && i>0 && i<ritz_vectors_.cols()-1) {
+        ritz_vectors_.col(i+1) = perform_op(i+1);
+      }
+    }
+    return num_converge;
+  }
+
+  size_t num_iterations() {
+    return m_niter_;
+  }
+
+  void sort_eigenpair() {
```

Comment:
I'm not sure why this function is necessary, since the power method only ever computes a single eigenpair ...

Reply:
Hmmm. I actually would like 3 eigenpairs, will gtell you why later. @jingwuOUO  can it actually do this?

Reply:
@dellaert the (optionally accelerated) power method only computes a **single** eigenpair estimate.  

It *is* possible to estimate k > 1 eigenpairs using the power method as a *subroutine* through a process called *deflation*, but this involves performing k sequential runs of the power method to compute the eigenpairs one by one.  If you really do need multiple eigenpairs, a proper Krylov subspace method like the Lanczos method really is the way to go.

Reply:
Yeah, that was my understanding too. But @jingwuOUO why do we then have *multiple* pairs in this code?

Reply:
That was my misunderstanding.  We actually do not need this sort_eigenpair(). Power method is returning only one eigenpair.

If we want to implement lanczos method, this can be used in the future.